skimage/filters/_gabor.py

```diff
@@ -13,11 +15,187 @@ def _sigma_prefactor(bandwidth):
+    sigma : scalar or vector, optional
+        Spatial dimensions of the envelope. Defaults to 1.
+    ndim : int, optional
+        Dimensionality of the envelope to create. Defaults to 2.
+
+    Returns
+    -------
+    gauss : (``ndim``, ``ndim``)
+
+    References
+    ----------
+    .. [1] Do CB. 2008. The Multivariate Gaussian Distribution. Stanford
+           University (CS 229): Stanford, CA.
+           http://cs229.stanford.edu/section/gaussians.pdf
+    """
+    sigma_prod = np.prod(sigma)
+
+    # normalization factor
+    norm = (2 * np.pi) ** (ndim / 2)
+    norm *= sigma_prod
+
+    # center image
+    image = image - center
+
+    # gaussian envelope
+    gauss = np.exp(-0.5 * np.sum(image ** 2) / sigma_prod ** 2)
+
+    return gauss / norm
+
+
+def _rotation(src_axis, dst_axis):
+    """Generates an matrix for the nD rotation of one axis to face another.
+
+    Parameters
+    ----------
+    src_axis : (N,) matrix
+        Vector representation of the axis that will be rotated.
+
+    dst_axis : (N,) matrix
+        Vector representation of the axis to rotate to.
+
+    Returns
+    -------
+    M : (N, N) array
+        Matrix that rotates ``src_axis`` to coincide with ``dst_axis``.
+
+    References
+    ----------
+    .. [1] Ognyan Ivanov Zhelezov. One Modification which Increases Performance
+           of N-Dimensional Rotation Matrix Generation Algorithm. International
+           Journal of Chemistry, Mathematics, and Physics, Vol. 2 No. 2, 2018:
+           pp. 13-18. https://dx.doi.org/10.22161/ijcmp.2.2.1
+
+    Examples
+    --------
+    >>> X = np.asarray([1, 0])
+    >>> Y = np.asarray([0.5, 0.5])
+
+    >>> M = _rotation(X, Y)
+    >>> Z = np.matmul(M, Y[..., np.newaxis])
+
+    >>> np.allclose(Z, Y[np.newaxis])
+    True
+    """
+    X = np.array(src_axis)
+    Y = np.array(dst_axis)
+
+
+    def rot(axis, diffs):
+        ndim = len(axis)
+        num_diffs = len(diffs)
+
+        x = axis
+        w = diffs
+
+        R = np.eye(ndim)  # Initial rotation matrix = Identity matrix
+
+        step = 1  # Initial step
+        while step < ndim:  # Loop to create matrices of stages
+            A = np.eye(ndim)
+
+            n = 0
+            while n < ndim - step and n + step < num_diffs:
+                r2 = x[w[n]] * x[w[n]] + x[w[n + step]] * x[w[n + step]]
+                if r2 > 0:
+                    r = np.sqrt(r2)
+                    pcos = x[w[n]] / r  # Calculation of coefficients
+                    psin = -x[w[n + step]] / r
+                    # Base 2-dimensional rotation
+                    A[w[n], w[n]] = pcos
+                    A[w[n], w[n + step]] = -psin
+                    A[w[n + step], w[n]] = psin
+                    A[w[n + step], w[n + step]] = pcos
+                    x[w[n + step]] = 0
+                    x[w[n]] = r
+
+                n += step << 1  # Move to the next base operation
+
+            step <<= 1  # multiply by 2
```

Comment:
This might be faster, but I don't think it matters enough here to obfuscate over the obvious: `step *= 2`.

In fact, this whole loop should be rewritten as `for step in 2 ** np.arange(np.log2(ndim))`. (For-loops in general should be strongly preferred to while loops because it keeps the start and end conditions close to each other, which massively helps readability.)

The inner loop could similarly be rewritten: `for n in range(0, ndim - step, step * 2)`. (with a `break` statement for `n + step > num_diffs`).

Reply:
@jni 
1) good point; the benefits are minimal
2) I disagree here because of the potential wasted RAM from the use of `2 ** np.arange` which unnecessarily calculates all of the steps at once which can be very costly if `ndim` reaches a high enough number. I prefer a while loop here because it effectively counts as lazy/step-by-step evaluation.
3) I'd be up to do this, but I'd need to use `xrange` to ensure lazy evaluation for both Python 2 and 3.

Reply:
@kne42 the cross-version idiom for lazy range is 

```python
from six import range  # xrange in Py2, range in Py3
```

You can of course do your own log-range call:

```
steps = (2**i for i in range(ceil(np.log2(ndim)))
for step in steps:
```

However at this point it's getting pretty complicated.

On the other hand, `np.log2(ndim)` is 30 for `ndim ≈ 1e9`. I find it hard to imagine that someone would want a 1,000,000,000-dimensional rotation matrix and then be worried about a size-30 array allocation? Am I missing something?